source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -50,11 +54,130 @@ void Filter::onRead() {
 
 void Filter::onReadWorker() {
   Network::ConnectionSocket& socket = cb_->socket();
-  std::string proxy_line;
-  if (!readLine(socket.fd(), proxy_line)) {
+
+  if ((!proxyProtocolHeader_.has_value() && !readProxyHeader(socket.fd())) ||
```

Comment:
This if statement is now a lot more complicated and it's not super clear what is going on with error checking, early return, etc. More comments? In what cases do we early return?

Reply:
done